youtube_dl/extractor/adn.py

```diff
@@ -119,59 +121,71 @@ def _get_subtitles(self, sub_path, video_id):
-                'index.php?option=com_vodapi&task=subtitles.getJSON&format=json&id=' + video_id
-            sub_path += '&token=' + token
-            error = links_data.get('error')
-        title = metas.get('title') or video_info['title']
+        config_url = 'https://gw.api.animedigitalnetwork.fr/player/video/%s/configuration' % video_id
+        player_config = self._download_json(
+            config_url, video_id,
+            'Downloading player config JSON metadata')['player']
+
+        options = player_config['options']
+        if not options['user'].get('hasAccess'):
+            raise ExtractorError('This video is only available for paying users')
+            # self.raise_login_required() # FIXME: Login is not implemented
+        tokens = self._download_json('https://gw.api.animedigitalnetwork.fr/player/refresh/token', video_id, 'Downloading access token', headers={'x-player-refresh-token': options['user']['refreshToken']}, data={})
+
+        links_url = player_config.get('linksurl') or options.get('videoUrl') or options['video']['url']
+        token = tokens['token']
+        self._K = ''.join([random.choice('0123456789abcdef') for _ in range(16)])
+        message = bytes_to_intlist(json.dumps({
+            'k': self._K,
+            't': token,
+        }))
+        links_data = None
+
+        # Sometimes authentication fails for no good reason, retry with
+        # a different random padding
+        login_attempt = 1
+        while login_attempt <= 3:
+            try:
+                padded_message = intlist_to_bytes(pkcs1pad(message, 128))
+                n, e = self._RSA_KEY
+                encrypted_message = long_to_bytes(pow(bytes_to_long(padded_message), e, n))
+                authorization = base64.b64encode(encrypted_message).decode()
+
+                links_data = self._download_json(
+                    urljoin(self._BASE_URL, links_url), video_id,
+                    'Downloading links JSON metadata', headers={
+                        'X-Player-Token': authorization
+                    },
+                    query={
+                        'freeWithAds': 'true',
+                        'adaptive': 'false',
+                        'withMetadata': 'true',
+                        'source': 'Web'
+                    }, expected_status=[200, 403]
+                )
+            except ExtractorError as e:
+                if isinstance(e.cause, compat_HTTPError) and e.cause.code == 401:
+                    login_attempt += 1
+                    self.report_warning('Authentication failed')
+                    self._sleep(1, None, msg_template='Waiting for %(timeout)s seconds before trying again')
+                else:
+                    raise e
+            else:
+                break
+
+        if links_data is None:
+            raise ExtractorError('Giving up retrying')
+
+        if links_data.get('code') == 'player-bad-geolocation-country':
```

Comment:
add an `only_matching` test for this case.

Reply:
Afaik everything's not available in Germany. Definitely the existing _TEST.